proto/buf/alpha/audit/v1alpha1/event.proto

```diff
@@ -73,5 +74,13 @@ message Event {
   // Time of the audited event. It specifies "when" it happened.
   google.protobuf.Timestamp event_time = 5;
   // Payload of the audited event. It specifies additional context on the event.
-  // TODO add oneof payload events.
+  oneof payload {
```

Comment:
I'd prefer to [stay away from nested messages](https://docs.buf.build/best-practices/style-guide#recommendations) if we can. Un-nesting, I suppose we could do:

```proto
message EventMetadata {
  string event_id
  EventType type
  Actor actor
  Resource resource
  google.protobuf.Timestamp event_time
}

message OrganizationCreatedEvent {
  EventMetadata metadata
  
  string name
}

// ... other events in a similar vein

message Event {
  oneof event {
    OrganizationCreatedEvent organization_created
  }
}
```

I think I like the original approach a little better because it reduces the nesting to the common fields, which I think cover most of what we're looking to capture.

Reply:
I prefer too the original approach. The advantage from both of these approaches is that we could force event emitters to send exactly the payload we expect. But to be honest, we could do that also with the original approach, receiving common fields as separated params, and the `Payload<Event>` as the last param, and the logger builds it back into an `AuditEvent`. I thought we agreed on keeping the original approach for now.

Reply:
> I'd prefer to [stay away from nested messages](https://docs.buf.build/best-practices/style-guide#recommendations) if we can.

the nested message definitions in my example weren't really related to the point I was trying to make, if that was the hangup -- sorry.

```
message OrganizationCreatedEvent {
  EventMetadata metadata
  
  string name
}

message Event {
  oneof payload {
    OrganizationCreatedEvent organization_created
  }
}
```

going with ^, I was mostly wondering about our preference in access pattern for fields

the original:
- accessing common metadata: `event.`
- accessing event-specific data: `event.payload.organization_created.`

vs. proposed:
- accessing common metadata: `event.payload.metadata.`
- accessing event-specific data: `event.payload.organization_created.`

the proposed puts metadata and "additional context" at the same "level" in the field hierarchy.

> I thought we agreed on keeping the original approach for now.

@unmultimedio for sure; I'm just thinking out in the open. Let me know if you feel like it's becoming too noisy or otherwise not adding value.

Reply:
@njhale sure, let's keep throwing ideas here!

So what I see from whatever approach we take we should make sure to:
- Have a single type that includes it all for the `rpc ListAuditEvents() []AuditEvent`.
- Force type safety for loggers, to avoid mismatch in the `event_type` and the actual `payload.<SomeEvent>`.

With the current approach, I think we could accomplish both. We already have:

```proto
message Event {
  EventType event_type = 1;
  EventActor actor = 2;
  EventResource resource = 3;
  Time event_time = 4;
  oneof payload {
    PayloadOrganizationCreated organization_created = 5;
    // ...
  }
}
```

And the callers currently do:

```go
ev := audit.Event{
  ...
  // the risk is we could have a mismatch here
  EventType: audit.ORG_CREATED,
  OrganizationCreated: audit.PayloadOrganizationCreated{...},
}
s.auditLogger(ctx, ev)
```

So we could force them with many interfaces just as you proposed a few days ago, similar to:

```go
type OrganizationLogger interface {
  func LogOrganizationCreated(
    context.Context,
    // all the commons (see there's no EventType, it's implied and it would be hydrated in the logger)
    audit.Actor,
    audit.Resource,
    // then the specific payload
    audit.PayloadOrganizationCreated,
  )
  // ... other func LogOrganization<OtherEvents>()
}

// ... other interfaces, one per service?
```

and the caller won't have anymore the possibility to mix things up:

```go
payload := audit.PayloadOrganizationCreated{...}
s.organizationLogger.LogOrganizationCreated(ctx, actor, resource, payload)
```

wdyt?

Reply:
could we do:

```go

  func Log(
    context.Context,
    audit.Actor,
    audit.Resource,
    audit.Payload, // interface
  )
```

a little more difficult for callers but makes it harder to make mistakes? I'd prefer not to add a bunch of separate interfaces

Reply:
Looking at the binary/json representation I think we should stick with the metadata in the envelope rather than in the oneof payloads:
```
{"organization_created": {"metadata": {"id": ..., "actor": ...}}, "name": "foo"}}
{"organization_deleted": {"metadata": {"id": ..., "actor": ...}}, "name": "foo"}}
```
vs
```
{"id": ..., "actor": ..., "type": "ORG_CREATED", "organization_created": {"name": "foo", ...}}
{"id": ..., "actor": ..., "type": "ORG_DELETED", "organization_deleted": {"name": "foo", ...}}
```
I think the bottom is easier to grep/jq/groq through if they wanted a text dump.

Reply:
> a little more difficult for callers but makes it harder to make mistakes? I'd prefer not to add a bunch of separate interfaces

@stefanvanburen you mean like the current catch-all interface we have now? I don't see how we can force the correct payload from the caller side with that interface, do you have something in mind?